src/dbnode/persist/fs/files.go

```diff
@@ -373,51 +415,136 @@ func (a fileSetFilesByTimeAndVolumeIndexAscending) Less(i, j int) bool {
 	return ti.Equal(tj) && ii < ij
 }
 
-func componentsAndTimeFromFileName(fname string) ([]string, time.Time, error) {
-	components := strings.Split(filepath.Base(fname), separator)
-	if len(components) < 3 {
-		return nil, timeZero, fmt.Errorf("unexpected file name %s", fname)
+// Returns the positions of filename delimiters ('-' and '.') and the number of
+// delimeters found, to be used in conjunction with the intComponentFromFilename
+// function to extract filename components. This function is deliberately
+// optimized for speed and lack of allocations, since filename parsing is known
+// to account for a significant proportion of system resources.
```

Comment:
"since allocation-heavy filename parsing can quickly become the large source of allocations in the entire system, especially when namespaces with long retentions are configured."